contracts/staking/GoodFundManager.sol

```diff
@@ -290,66 +298,62 @@ contract GoodFundManager is DAOUpgradeableContract, DSMath {
 	 */
-	function calcSortedContracts(uint256 _maxGasAmount)
-		public
-		view
-		returns (address[] memory)
-	{
-		uint256 activeContractsLength = activeContracts.length;
-		address[] memory addresses = new address[](activeContractsLength);
-		uint256[] memory balances = new uint256[](activeContractsLength);
+	function calcSortedContracts() public view returns (InterestInfo[] memory) {
+		address[] memory addresses = new address[](activeContracts.length);
+		uint256[] memory balances = new uint256[](activeContracts.length);
+		InterestInfo[] memory interestInfos = new InterestInfo[](
+			activeContracts.length
+		);
 		uint256 tempInterest;
 		int256 i;
-		for (i = 0; i < int256(activeContractsLength); i++) {
+		for (i = 0; i < int256(activeContracts.length); i++) {
 			(, , , , tempInterest) = IGoodStaking(activeContracts[uint256(i)])
 				.currentGains(false, true);
 			if (tempInterest != 0) {
 				addresses[uint256(i)] = activeContracts[uint256(i)];
 				balances[uint256(i)] = tempInterest;
 			}
 		}
-		uint256 gasCostInDAI = getGasPriceIncDAIorDAI(_maxGasAmount, true); // Get gas price in DAI so can compare with possible interest amount to get
-		address[] memory emptyArray = new address[](0);
-
+		uint256 usedGasAmount = gasCostExceptInterestCollect;
 		quick(balances, addresses); // sort the values according to interest balance
 		uint256 gasCost;
 		uint256 possibleCollected;
-		for (i = int256(activeContractsLength) - 1; i >= 0; i--) {
+		uint256 maxGasAmount;
+		for (i = int256(activeContracts.length) - 1; i >= 0; i--) {
 			// elements are sorted by balances from lowest to highest
 
 			if (addresses[uint256(i)] != address(0x0)) {
 				gasCost = IGoodStaking(addresses[uint256(i)])
 					.getGasCostForInterestTransfer();
-				if (_maxGasAmount - gasCost >= gasCostExceptInterestCollect) {
-					// collects the interest from the staking contract and transfer it directly to the reserve contract
-					//`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
-					possibleCollected += balances[uint256(i)];
-					_maxGasAmount = _maxGasAmount - gasCost;
-				} else {
-					break;
-				}
+
+				// collects the interest from the staking contract and transfer it directly to the reserve contract
+				//`collectUBIInterest` returns (iTokengains, tokengains, precission loss, donation ratio)
+				possibleCollected += balances[uint256(i)];
+				usedGasAmount += gasCost;
+				maxGasAmount = block.timestamp >=
+					lastCollectedInterest + collectInterestTimeThreshold
+					? (possibleCollected * 1e10) /
+						getGasPriceIncDAIorDAI(1, true)
+					: (possibleCollected * 1e10) /
+						(interestMultiplier * getGasPriceIncDAIorDAI(1, true));
+				interestInfos[uint256(i)] = InterestInfo({
+					contractAddress: addresses[uint256(i)],
+					interestBalance: balances[uint256(i)],
+					collectedInterestSoFar: possibleCollected,
+					gasCostSoFar: usedGasAmount,
```

Comment:
what about the "fixed" gas costs

Reply:
It's included I set usedGasAmount initially to fixed gas costs